apps/astarte_realm_management/lib/astarte_realm_management/engine.ex

```diff
@@ -273,43 +273,89 @@ defmodule Astarte.RealmManagement.Engine do
     end
   end
 
-  # TODO: Mappings documentation changes are discarded
   defp extract_new_mappings(db_client, %{mappings: upd_mappings} = interface_doc) do
     descriptor = InterfaceDescriptor.from_interface(interface_doc)
 
-    with {:ok, mappings} <- Mappings.fetch_interface_mappings(db_client, descriptor.interface_id) do
+    with {:ok, mappings_map} <-
+           Mappings.fetch_interface_mappings_map(db_client, descriptor.interface_id,
+             include_docs: true
+           ) do
       upd_mappings_map =
         Enum.into(upd_mappings, %{}, fn mapping ->
           {mapping.endpoint_id, mapping}
         end)
 
-      maybe_new_mappings =
-        Enum.reduce_while(mappings, upd_mappings_map, fn mapping, acc ->
-          case drop_mapping_doc(Map.get(upd_mappings_map, mapping.endpoint_id)) do
-            nil ->
-              {:halt, {:error, :missing_endpoints}}
+      fetch_new_mappings(upd_mappings_map, mappings_map)
+    end
+  end
 
-            ^mapping ->
-              {:cont, Map.delete(acc, mapping.endpoint_id)}
+  defp fetch_new_mappings(upd_mappings_map, mappings_map) do
+    keys = Map.keys(mappings_map)
+    {maybe_new_mappings, new_mappings} = Map.split(upd_mappings_map, keys)
+
+    if Enum.count(upd_mappings_map) < Enum.count(mappings_map) do
+      {:error, :missing_endpoints}
+    else
+      updated_mappings =
+        Enum.reduce_while(mappings_map, maybe_new_mappings, fn {mapping_id, mapping}, acc ->
+          with {:ok, upd_mapping} <- Map.fetch(maybe_new_mappings, mapping_id),
+               :ok <- check_valid_mapping_update(upd_mapping, mapping) do
+            {:cont, acc}
+          else
+            :error ->
+              {:cont, acc}
 
-            _ ->
+            {:error, :incompatible_endpoint_change} ->
               {:halt, {:error, :incompatible_endpoint_change}}
+
+            {:error, :no_applicable_change} ->
+              {:cont, Map.delete(acc, mapping_id)}
           end
         end)
```

Comment:
I'd move this in a separate function like
```elixir
defp extract_updated_mappings(existing_mappings, old_mappings) do
  Enum.reduce_while(old_mappings, {:ok, %{}}, fn {mapping_id, old_mapping}, {:ok, acc} ->
    with {:ok, updated_mapping} <- Map.fetch(existing_mappings, mapping_id),
         {:allowed, true} <- {:allowed, allowed_mapping_update?(updated_mapping, old_mapping)},
         {:updated, true} <- {:updated, is_mapping_updated?(updated_mapping, old_mapping)} do
      {:cont, {:ok, Map.put(acc, mapping_id, updated_mapping)}
    else
      :error ->
        {:halt, {:error, :missing_endpoints}}
  
      {:allowed, false} ->
        {:halt, {:error, :incompatible_endpoint_change}}
  
      {:updated, true} ->
        {:cont, {:ok, acc}}
    end
  end)
end
```
You can drop the `Enum.count` check above so you don't go through the maps twice, and writing the function that way means that you can just remove the `fetch_new_mappings` function, move the 2 lines of its implementation into `extract_new_mappings` and finish it with
```elixir
with {:ok, updated_mappings} <- extract_updated_mappings do
  {:ok, Map.merge(new_mappings, updated_mappings)}
end
```